functional-tests/appwrapper_reconciler_test.go

```diff
@@ -0,0 +1,229 @@
+	log "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+func logger() {
+	// Get log file path from environment variable or use a default path
+	logFilePath := os.Getenv("LOG_FILE_PATH")
+	if logFilePath == "" {
+		logFilePath = "./functional-test-logfile.log"
+	}
+	// Create a log file
+	logFile, err := os.Create(logFilePath)
+	if err != nil {
+		log.Log.Error(err, "Error creating log file: %v", err)
+	}
+	// Configure zap logger to write to the log file
+	logger := zap.New(zap.WriteTo(logFile), zap.UseDevMode(true))
+
+	// Set the logger for controller-runtime
+	ctrl.SetLogger(logger)
+
+	// This line prevents controller-runtime from complaining about log.SetLogger never being called
+	log.SetLogger(logger)
+}
+
+func startEnvTest(t *testing.T) *rest.Config {
+	// to redirect all functional test related logs to separate logfile ~ default (in functional-tests directory), can be changed using environment variable LOG_FILE_PATH=/path_to_logfile
+	logger()
+
+	//specify testEnv configuration
+	testEnv := &envtest.Environment{
+		CRDDirectoryPaths: []string{
+			filepath.Join(build.Default.GOPATH, "pkg", "mod", "github.com", "project-codeflare", "multi-cluster-app-dispatcher@v1.39.0", "config", "crd", "bases"),
+			filepath.Join(build.Default.GOPATH, "pkg", "mod", "github.com", "openshift", "api@v0.0.0-20230213134911-7ba313770556", "machine", "v1beta1"),
+		},
+		ErrorIfCRDPathMissing: true,
+	}
+	test := WithConfig(t, testEnv.Config)
+	cfg, err := testEnv.Start()
+	test.Expect(err).NotTo(HaveOccurred())
+
+	t.Cleanup(func() {
+		teardownTestEnv(t, testEnv)
+	})
+
+	return cfg
+}
+
+func startInstascaleController(test Test, cfg *rest.Config) {
+	// Add custom resource schemes to the global scheme.
+	err := mcadv1beta1.AddToScheme(scheme.Scheme)
+	test.Expect(err).NotTo(HaveOccurred())
+
+	err = machinev1beta1.AddToScheme(scheme.Scheme)
+	test.Expect(err).NotTo(HaveOccurred())
+
+	// Create a controller manager for managing controllers.
+	k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
+		Scheme: scheme.Scheme,
+	})
+	test.Expect(err).ToNot(HaveOccurred())
+
+	// Create an instance of the AppWrapperReconciler with configuration
+	instaScaleController := &controllers.AppWrapperReconciler{
+		Client: k8sManager.GetClient(),
+		Scheme: k8sManager.GetScheme(),
+		Config: config.InstaScaleConfiguration{
+			MachineSetsStrategy: "reuse",
+			MaxScaleoutAllowed:  5,
+		},
+	}
+	// Set up the AppWrapperReconciler with the manager.
+	err = instaScaleController.SetupWithManager(context.Background(), k8sManager)
+	test.Expect(err).ToNot(HaveOccurred())
+
+	// Start the controller manager in a goroutine.
+	go func() {
+		err = k8sManager.Start(ctrl.SetupSignalHandler())
+		test.Expect(err).ToNot(HaveOccurred())
+	}()
+
+}
+
+func teardownTestEnv(t *testing.T, testEnv *envtest.Environment) {
+	if err := testEnv.Stop(); err != nil {
+		t.Log("Error stopping test Environment\n", err)
```

Comment:
Hey :)
This error is getting logged during teardown, 

```
=== RUN   TestReconciler
    test.go:132: Creating ephemeral output directory as CODEFLARE_TEST_OUTPUT_DIR env variable is unset
    test.go:135: Output directory has been created at: /tmp/TestReconciler2973227348/001
    appwrapper_reconciler_test.go:110: Error stopping test Environment
         timeout waiting for process kube-apiserver to stop
--- PASS: TestReconciler (27.03s)
PASS
ok  	github.com/project-codeflare/instascale/functional-tests
```

I see there was a similar issue raised here: https://github.com/kubernetes-sigs/controller-runtime/issues/1571

Reply:
Yes, I tried to nullify this error using above mentioned reference, also considered adjusting timeout duration but couldn't investigate root cause of the delay in stopping the test environment process.

Reply:
This error is resolved.